plasmapy/diagnostics/thomson.py

```diff
@@ -319,3 +345,207 @@ def spectral_density(
+    fit,
+    efract: np.ndarray = None,
+    ifract: np.ndarray = None,
+    ion_species: Union[str, List[str], Particle, List[Particle]] = "H+",
+    probe_vec=np.array([1, 0, 0]),
+    scatter_vec=np.array([0, 1, 0]),
+    n: u.m ** -3 = None,
+    Te: u.K = None,
+    Ti: u.K = None,
+    electron_vel: u.m / u.s = None,
+    ion_vel: u.m / u.s = None,
+):
+    r"""
+    Function that fits Thomson scattering spectra.
+    """
+
+    # Condition the raw input in the same way as the thomson.spectral_density
+    # function does
+    (
+        Te,
+        Ti,
+        efract,
+        ifract,
+        ion_species,
+        electron_vel,
+        ion_vel,
+    ) = _condition_thomson_input(
+        Te, Ti, efract, ifract, ion_species, electron_vel, ion_vel
+    )
+    # Extract the number of ion and electron populations
+    num_i = len(ifract)
+    num_e = len(efract)
+
+    settings = {
+        "ion_species": ion_species,
+        "probe_wavelength": probe_wavelength,
+        "probe_vec": probe_vec,
+        "scatter_vec": scatter_vec,
+        "num_i": num_i,
+        "num_e": num_e,
+    }
+
+    params = lmfit.Parameters()
+    params.add("n", value=n.to(u.cm ** -3).value, vary=False, min=0)
+
+    # Break apart vectorized inputs into individual quantities
+    # This is necessary because lmfit doesn't support vector parameters
+    for i in range(num_e):
+        params.add("efract" + str(i), value=efract[i], vary=False, min=0, max=1)
+        params.add("Te" + str(i), value=Te[i].to(u.K).value, vary=False, min=0)
+        for j in range(3):
+            params.add(
+                f"ve{i}_{j}", value=electron_vel[i, j].to(u.m / u.s).value, vary=False
+            )
+
+    for i in range(num_i):
+        params.add("ifract" + str(i), value=ifract[i], vary=False, min=0, max=1)
+        params.add("Ti" + str(i), value=Ti[i].to(u.K).value, vary=False, min=0)
+        for j in range(3):
+            params.add(
+                f"vi{i}_{j}", value=ion_vel[i, j].to(u.m / u.s).value, vary=False
+            )
+
+    # Catch references to single varaibles and append a 0
+    for v in fit:
+        if v in ["efract", "Te"] and num_e == 1:
+            fit.append(v + "0")
+            fit.remove(v)
+        elif v in ["ifract", "Ti"] and num_i == 1:
+            fit.append(v + "0")
+            fit.remove(v)
+
+    for v in fit:
+        valid_keys = list(params.keys())
+        if v in valid_keys:
+            params[v].vary = True
+        else:
+            raise KeyError(
+                f"Fit variable {v} is not valid for the provided "
+                f"fit parameters: {valid_keys}"
+            )
+
+    # params.pretty_print()
+
+    fit_fcn = lambda params, wavelengths, Skw: _spectral_density_fit_function(
+        params, wavelengths, Skw, settings
+    )
+
+    minner = lmfit.Minimizer(
+        fit_fcn,
+        params,
+        fcn_args=(wavelengths.to(u.nm).value, spectral_density.to(u.s / u.rad).value),
+    )
```

Comment:
I ran the test through a debugger, and got this kind of result:

```
<Parameter 'n', value=4.999999999999992e+17 +/- 42.5, bounds=[0:inf]>
<Parameter 'Te0', value=116045.18121550081 (fixed), bounds=[0:inf]> 
```

which, I guess, answers my earlier question from looking through the notebook. This is, indeed, a cool estimate - but I worry that the error bars may be getting underappreciated. I suppose that this, being simply a fitting routine, should not carry too much additional error, but it is something I worry about.

Reply:
I think you're right - one of my action items (now added to the post at the top) is to somehow estimate the fit error bars at the end, possibly by systematically varying the best-fit parameters and seeing how that changes Chi2.

Reply:
So, it looks like estimating fit errors for such a complicated, non-linear function is hard! Right now I'm using `lmfit.conf_interval` as one approach. That works approximately, although it does still under-estimate the error in most cases (sometimes worse than others). 

`lmfit` also includes a wrapper for the Bayesian `emcee` MCMC model, so I've tried doing that as well. It works great, except for two issues. 1) The distributions often still seem to under estimate the error (the actual value is not within the likely region of the posterior) and 2) it takes almost an hour to run one fit on 1 CPU core! 

![corner_plot](https://user-images.githubusercontent.com/32618747/107529134-8ff73800-6b88-11eb-9446-70c0d77c6710.png)